rust/connlib/tunnel/src/dns.rs

```diff
@@ -286,39 +261,61 @@ impl StubResolver {
         }
 
         let question = message.first_question()?;
+        let domain = question.qname().to_vec();
+        let qtype = question.qtype();
 
-        tracing::trace!("Parsed packet as DNS query: '{question}'");
+        tracing::trace!("Parsed packet as DNS query: '{qtype} {domain}'");
 
-        if let Some(records) = self.known_hosts.get_records(&question) {
-            let response =
-                build_dns_with_answer(message, question.qname(), question.qtype(), records)?;
+        if let Some(records) = self.known_hosts.get_records(qtype, &domain) {
+            let response = build_dns_with_answer(message, domain, records)?;
             return Some(ResolveStrategy::LocalResponse(build_response(
                 packet, response,
             )));
         }
 
-        if !self.is_resource(&question) {
-            return Some(ResolveStrategy::ForwardQuery(DnsQuery {
-                name: ToName::to_vec(question.qname()),
-                record_type: u16::from(question.qtype()).into(),
-                query: packet,
-            }));
-        }
+        let resource_records = match qtype {
+            Rtype::A if self.is_fqdn_resource(&domain) => {
+                self.get_or_assign_a_records(domain.clone())
+            }
+            Rtype::AAAA if self.is_fqdn_resource(&domain) => {
+                self.get_or_assign_aaaa_records(domain.clone())
+            }
+            Rtype::PTR => {
+                let fqdn = self.resource_address_name_by_reservse_dns(&domain)?;
 
-        let resource_records = self.get_records(&question);
+                vec![AllRecordData::Ptr(domain::rdata::Ptr::new(fqdn))]
+            }
+            _ => {
+                return Some(ResolveStrategy::ForwardQuery(DnsQuery {
+                    name: domain,
+                    record_type: u16::from(qtype).into(),
+                    query: packet,
+                }))
+            }
+        };
```

Comment:
@conectado Let me know what you think of this. IMO having more of the decision logic in one place here makes things easier to understand, even if there is a slight duplication with the `self.is_fqdn_resource` call.